### DISCUSSION TITLE
References for FFI and calling C from V

### DISCUSSION THREAD

**Discussion by KevinDaSilvaS** [vlang/v#16659] [Questions and Answers]
Hey everyone I was trying to use C libraries in my V code, but I didnt manage to make it work on windows using the docs available(probably because I dont know C), someone has  access to more detailed documentation on it?

**Comment by JalonSolov** [comment]
For now, you definitely need to know C to get this to work.

Perhaps in the future, V will get a WinAPI wrapper layer to work with the system, and other wrapper layers for other useful Windows versions of C libraries.

You can try running `v translate wrapper` on some of those C library files, but if it doesn't work, you won't be able to figure out why, nor be able to write your own wrappers, without knowing C.

**Comment by Wajinn** [comment]
Agree that the best case will likely be to translate C to V, else the person should be familiar with or studying some C too, to better understand what's possibly going wrong or needs changes.  

Depending on what a person is doing on Windows, if using the default TCC, you might need the full winapi. 
TCC's full winapi (named as- winapi-full-for-0.9.27.zip) can be found [here](https://download-mirror.savannah.gnu.org/releases/tinycc/) . Download and extract to somewhere that doesn't include spaces (to avoid issues). 

You can then do an alternative run/compile with the below (example).
`v -cc tcc -cflags "-isystem path\to\full-winapi\include\winapi" run example.v`

In regards to the V documentation, the [example](https://github.com/vlang/v/blob/master/doc/docs.md#calling-c-from-v) does kind of show how it's done, but is not a simple example for someone to start out with.  I suppose a bit more or another example could be added to the documentation. 

```
/* 
 * May need to use #flag and/or #include at top, depending on C functions used. 
 * On Windows, may need full winapi for TCC, as header/function found there.
 * https://github.com/vlang/v/blob/master/doc/docs.md#passing-c-compilation-flags
 * https://github.com/vlang/v/blob/master/doc/docs.md#including-c-code
 * Or attemept to translate C to V - https://github.com/vlang/v/blob/master/doc/docs.md#translating-c-to-v
 * Simple example below
*/

fn  C.GetSystemMetrics(nIndex int) int // Declare C functions that are to be used.

fn main() {
	height, width := screen_size()  
	println('${height}') 
	println('${width}')  

}

fn screen_size() (int, int) {
	height := int(C.GetSystemMetrics(C.SM_CYSCREEN))    
	width := int(C.GetSystemMetrics(C.SM_CXSCREEN))  
	return height, width  
}
```

**Comment by terremoth** [comment]
Will be nice to have support for FFI and Win32 API, also [COM](https://en.wikipedia.org/wiki/Component_Object_Model) access
